coreax/solvers/coresubset.py

```diff
@@ -458,3 +479,273 @@ def selection_function(i: int, _kernel_similarity_penalty: ArrayLike) -> Array:
+        coresets
+    :param responses: Array of responses
+    :param feature_gramian: Feature kernel gramian
+    :param regularisation_parameter: Regularisation parameter for stable inversion of
+        array, negative values will be converted to positive
+    :param identity: identity matrix
+    :param least_squares_solver: Instance of
+        :class:`coreax.least_squares.RegularisedLeastSquaresSolver`
+
+    :return: GreedyKernelInducingPoints loss for each candidate coreset
+    """
+    # Extract all the possible "next" coreset feature gramians, cross feature gramians
+    # and coreset response vectors.
+    coreset_gramians = feature_gramian[
+        (candidate_coresets[:, :, None], candidate_coresets[:, None, :])
+    ]
+    coreset_cross_gramians = feature_gramian[candidate_coresets, :-1]
+    coreset_responses = responses[candidate_coresets]
+
+    # Solve for the kernel ridge regression coefficients for each possible coreset
+    coefficients = least_squares_solver.solve_stack(
+        arrays=coreset_gramians,
+        regularisation_parameter=regularisation_parameter,
+        targets=coreset_responses,
+        identity=identity,
+    )
+
+    # Compute the loss function, making sure that we remove the padding on the responses
+    predictions = (coreset_cross_gramians * coefficients).sum(axis=1)
+    loss = ((predictions - 2 * responses[:-1, 0]) * predictions).sum(axis=1)
+    return loss
+
+
+# pylint: disable=too-many-locals
+class GreedyKernelInducingPoints(
+    RefinementSolver[_SupervisedData, GreedyKernelInducingPointsState],
+    ExplicitSizeSolver,
+):
+    r"""
+    Apply GreedyKernelInducingPoints to a supervised dataset.
+
+    GreedyKernelInducingPoints is a deterministic, iterative and greedy approach to
+    build a coreset.
+
+    Given one has an original dataset :math:`\mathcal{D}^{(1)} = \{(x_i, y_i)\}_{i=1}^n`
+    of ``n`` pairs with :math:`x\in\mathbb{R}^d` and :math:`y\in\mathbb{R}^p`, and one
+    has selected :math:`m` data pairs :math:`\mathcal{D}^{(2)} = \{(\tilde{x}_i,
+    \tilde{y}_i)\}_{i=1}^m` already for their compressed representation of the original
+    dataset, GreedyKernelInducingPoints selects the next point to minimise the loss
+
+    .. math::
+
+        L(\mathcal{D}^{(1)}, \mathcal{D}^{(2)}) = ||y^{(1)} -
+        K^{(12)}(K^{(22)} + \lambda I_m)^{-1}y^{(2)} ||^2_{\mathbb{R}^n}
+
+    where :math:`y^{(1)}\in\mathbb{R}^n` is the vector of responses from
+    :math:`\mathcal{D}^{(1)}`, :math:`y^{(2)}\in\mathbb{R}^n` is the vector of responses
+    from :math:`\mathcal{D}^{(2)}`,  :math:`K^{(12)} \in \mathbb{R}^{n\times m}`is the
+    cross-matrix of kernel evaluations between :math:`\mathcal{D}^{(1)}` and
+    :math:`\mathcal{D}^{(2)}`, :math:`K^{(22)} \in \mathbb{R}^{m\times m}` is the
+    kernel matrix on :math:`\mathcal{D}^{(2)}`,
+    :math:`\lambda I_m \in \mathbb{R}^{m \times m}`is the identity matrix and
```

Comment:
```suggestion
    :math:`\lambda I_m \in \mathbb{R}^{m \times m}` is the identity matrix and
```
space needed here also